skimage/exposure/_adapthist.py

```diff
@@ -162,64 +157,50 @@ def _clahe(image, kernel_size, clip_limit, nbins):
             clim = NR_OF_GRAY  # Large value, do not clip (AHE)
 
         hist = lut[sub_img.ravel()]
-        hist = np.bincount(hist)
+        hist = np.bincount(hist, minlength=nbins)
         hist = np.append(hist, np.zeros(nbins - hist.size, dtype=int))
         hist = clip_histogram(hist, clim)
         hist = map_histogram(hist, 0, NR_OF_GRAY - 1, sub_img.size)
         map_array[inds] = hist
 
-    # Interpolate greylevel mappings to get CLAHE image
-    offsets = [0] * image.ndim
-    lowers = [0] * image.ndim
-    uppers = [0] * image.ndim
-    starts = [0] * image.ndim
-    prev_inds = [0] * image.ndim
-
-    for inds in np.ndindex(*[ns[dim] + 1 for dim in range(image.ndim)]):
-
-        for dim in range(image.ndim):
-            if inds[dim] != prev_inds[dim]:
-                starts[dim] += offsets[dim]
-
-        for dim in range(image.ndim):
-            if dim < image.ndim - 1:
-                if inds[dim] != prev_inds[dim]:
-                    starts[dim + 1] = 0
-
-        prev_inds = inds[:]
-
-        # modify edges to handle special cases
-        for dim in range(image.ndim):
-            if inds[dim] == 0:
-                offsets[dim] = np.ceil(steps[dim] / 2.0)
-                lowers[dim] = 0
-                uppers[dim] = 0
-            elif inds[dim] == ns[dim]:
-                offsets[dim] = np.ceil(steps[dim] / 2.0)
-                lowers[dim] = ns[dim] - 1
-                uppers[dim] = ns[dim] - 1
-            else:
-                offsets[dim] = steps[dim]
-                lowers[dim] = inds[dim] - 1
-                uppers[dim] = inds[dim]
-
-        maps = []
-        for edge in np.ndindex(*([2] * image.ndim)):
-            maps.append(map_array[tuple([[lowers, uppers][edge[dim]][dim]
-                                         for dim in range(image.ndim)])])
-
-        slices = [np.arange(starts[dim],
-                            np.min([starts[dim] + offsets[dim],
-                                    image.shape[dim]]))
-                  for dim in range(image.ndim)]
-
-        interpolate(image, slices[::-1], maps, lut)
+    # Perform multilinear interpolation of graylevel mappings
+    # using the convention described here:
+    # https://en.wikipedia.org/w/index.php?title=Adaptive_histogram_
+    # equalization&oldid=936814673#Efficient_computation_by_interpolation
+
+    # determine num of blocks to be processed separately in each dim:
+    # generally n+1, n if last block only processes padded pixels
+    ns_proc = [n + 1 if (np.ceil(s / 2.) + (n - 1) * s) <= (sh - pe)
+               else n for sh, pe, s, n
+               in zip(image.shape, pad_end_per_dim, steps, ns)]
+
+    for inds in np.ndindex(*ns_proc):
+
+        # define slices for each dim
+        starts = [int(i > 0) * np.ceil(s / 2.) + np.max([0, i - 1]) * s
```

Comment:
Am I wrong or `i` can't be negative?
```suggestion
        starts = [np.ceil(s / 2.) + (i - 1) * s if i > 0 else 0
```

Reply:
True, the conditional statement is better than the awkward casting to `int` :)